src/frontend/Promotion.ml

```diff
@@ -61,7 +64,9 @@ let rec promote (exp : Ast.typed_expression) prom =
   | RowVectorExpr (_ :: _ as es) ->
       let pes = List.map ~f:(fun e -> promote e prom) es in
       let fst = List.hd_exn pes in
-      let ad_level = fst.emeta.ad_level in
+      let ad_level =
```

Comment:
This is the change that fixed the bug; the rest are just cleanups. I’m not sure why arrays previously seemed to work just fine while matrices didn’t, I couldn’t find where in the code they diverged in behavior.

Reply:
Arrays were broken too if the inner type is a row vector.
```stan
array[2] row_vector[2] Mar;
Mar = {[1, 2], [3, y]};
```